File: app/src/main/java/com/live/ui/juhe/BDLiveActivity.java
package com.live.ui.juhe;

import android.content.ComponentName;
import android.content.Intent;
import android.graphics.PixelFormat;
import android.media.AudioManager;
import android.os.Build;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.provider.Settings;
import android.view.MotionEvent;
import android.view.Window;
import android.view.WindowManager;

import com.baidu.cloud.media.player.IMediaPlayer;
import com.live.R;
import com.live.base.BaseActivity;
import com.live.utils.DeviceUtil;
import com.live.view.BDCloudVideo.BDCloudVideoView;
import com.live.view.ToastHelper;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.github.prototypez.savestate.core.annotation.AutoRestore;

public class BDLiveActivity extends BaseActivity implements IMediaPlayer.OnPreparedListener,
        IMediaPlayer.OnCompletionListener, IMediaPlayer.OnErrorListener,
        IMediaPlayer.OnInfoListener, IMediaPlayer.OnBufferingUpdateListener,
        BDCloudVideoView.OnPlayerStateListener {

    @BindView(R.id.vv_activty_bdlive)
    BDCloudVideoView mVV;

    @AutoRestore("url")
    String url;

    private int bright = 0;

    private AudioManager mAudioManager;

    private boolean isClickBack = false;
    //0 activity未启动 1 activity在后台 2 activity在前台
    public static int startBackground = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = this.getWindow();
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            // finally change the color
            window.setStatusBarColor(0xff282828);
        }
        //防闪屏
        getWindow().setFormat(PixelFormat.TRANSLUCENT);
        setContentView(R.layout.activity_bdlive);
        ButterKnife.bind(this);

        url = getIntent().getStringExtra("url");

        bright = getScreenBrightness();
//        //初始化音频管理器
        mAudioManager = (AudioManager) getSystemService(AUDIO_SERVICE);

        initPlayer();

        mVV.setVideoPath(url);
        // 初始化好之后立即播放（您也可以在onPrepared回调中调用该方法）
        if (!mVV.isPlaying()) {
            mVV.start();
        }
    }

    private float startY = 0;//手指按下时的Y坐标
    private float startX = 0;//手指按下时的X坐标


    private void initPlayer() {
        //设置ak
//        BDCloudVideoView.setAK("b98ce254325147ec83572bcf2abfc43d");
        //注册listener
        mVV.setOnPreparedListener(this);
        mVV.setOnCompletionListener(this);
        mVV.setOnErrorListener(this);
        mVV.setOnInfoListener(this);
        mVV.setOnBufferingUpdateListener(this);
        mVV.setOnPlayerStateListener(this);

        mVV.setVideoScalingMode(BDCloudVideoView.VIDEO_SCALING_MODE_SCALE_TO_FIT);

        mVV.setLogEnabled(false);
//        mVV.setDecodeMode(BDCloudMediaPlayer.DECODE_SW);
        mVV.setMaxProbeTime(2000); // 设置首次缓冲的最大时长
        mVV.setTimeoutInUs(1000000);
        // Options for live stream only
//        mVV.setMaxProbeSize(1 * 2048);
//        mVV.setMaxProbeTime(40); // 设置首次缓冲的最大时长
//        mVV.setMaxCacheSizeInBytes(32 * 1024);
//        mVV.setBufferTimeInMs(100);
//        mVV.toggleFrameChasing(true);
    }


    private int getScreenBrightness() {
        int screenBrightness = 255;
        try {
            screenBrightness = Settings.System.getInt(getContentResolver(), Settings.System.SCREEN_BRIGHTNESS);
        } catch (Exception ignored) {

        }
        return screenBrightness;
    }

    private void setLight(int brightness) {
        int temp = bright + brightness;
        if (temp < 0 || temp > 255) {
            return;
        }
        bright = temp;
        WindowManager.LayoutParams lp = getWindow().getAttributes();
        lp.screenBrightness = bright * (1f / 255f);
        getWindow().setAttributes(lp);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int screenWidth = mVV.getWidth();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                startX = event.getX();
                startY = event.getY();
                if (startY < DeviceUtil.dip2px(50)) {
                    break;
                }
                break;
            case MotionEvent.ACTION_MOVE:
                if (startY < DeviceUtil.dip2px(50)) {
                    break;
                }
                float endY = event.getY();
                float distanceY = startY - endY;
                if (startX > screenWidth / 2) {
                    //右边
                    //在这里处理音量
                    final double FLING_MIN_DISTANCE = DeviceUtil.dip2px(3);
                    if (distanceY > FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_RAISE, AudioManager.FX_FOCUS_NAVIGATION_UP);
                    }
                    if (distanceY < FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_LOWER, AudioManager.FX_FOCUS_NAVIGATION_UP);
                    }
                } else {
                    //屏幕左半部分上滑，亮度变大，下滑，亮度变小
                    final double FLING_MIN_DISTANCE = 10;
                    if (distanceY > FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        setLight(10);
                    }
                    if (distanceY < FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        setLight(-10);
                    }
                }
                startY = endY;
                break;
        }
        return super.onTouchEvent(event);
    }


    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
        if (mVV == null) {
            mVV = findViewById(R.id.vv_activty_bdlive);
            initPlayer();
        }
        String url = intent.getStringExtra("url");

        if (this.url.equals(url)) {
            mVV.enterForeground();
            this.url = url;
        } else {
            mVV.setVideoPath(url);
            mVV.start();
        }

    }

    @Override
    protected void onRestart() {
        super.onRestart();
        if (isClickBack) {
            return;
        }
        if (mVV != null) {
            mVV.enterForeground();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        isClickBack = false;
        startBackground = 2;
    }

    private CountDownTimer countDownTimer;

    @Override
    public void onBackPressed() {
        Intent intent = getIntent();
        if (intent == null) {
            intent = new Intent(this, AnchorListActivity.class);
        } else {
            ComponentName componentName = new ComponentName(this, AnchorListActivity.class);
            intent.setComponent(componentName);
        }
        startActivity(intent);
        moveTaskToBack(isTaskRoot());
        isClickBack = true;
        if (mVV != null) {
            mVV.stopPlayback();
            this.url = "";
            //后台60s自动关闭
            if (countDownTimer == null) {
                countDownTimer = new CountDownTimer(60000, 30000) {
                    @Override
                    public void onTick(long millisUntilFinished) {

                    }

                    @Override
                    public void onFinish() {
                        finish();
                    }
                };
            }
            countDownTimer.start();
        }
    }


    @Override
    protected void onStop() {
        // enterBackground should be invoke before super.onStop()
        if (isClickBack) {
            super.onStop();
            return;
        }
        if (mVV != null) {
            mVV.enterBackground();
        }
        startBackground = 1;
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        if (mVV != null) {
            mVV.stopPlayback(); // 释放播放器资源
            mVV.release(); // 释放播放器资源和显示资源
        }
        mVV = null;
        if (countDownTimer != null) {
            countDownTimer.cancel();
            countDownTimer = null;
        }
        startBackground = 0;
        super.onDestroy();
    }


    @Override
    public boolean onInfo(IMediaPlayer mp, int what, int extra) {
        return false;
    }

    @Override
    public boolean onError(IMediaPlayer mp, int what, int extra) {
        ToastHelper.showToast("出错啦出错啦出错啦");
        finish();
        return false;
    }

    @Override
    public void onCompletion(IMediaPlayer mp) {

    }

    @Override
    public void onPrepared(IMediaPlayer mp) {

    }

    @Override
    public void onBufferingUpdate(IMediaPlayer mp, int percent) {

    }

    @Override
    public void onPlayerStateChanged(BDCloudVideoView.PlayerState nowState) {

    }
}

File: app/src/main/java/com/live/ui/juhe/BDLiveActivity.kt
package com.live.ui.juhe

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.graphics.PixelFormat
import android.media.AudioManager
import android.os.Build
import android.os.Bundle
import android.os.CountDownTimer
import android.provider.Settings
import android.view.MotionEvent
import android.view.WindowManager
import com.baidu.cloud.media.player.IMediaPlayer
import com.live.R
import com.live.base.BaseActivity
import com.live.utils.DeviceUtil
import com.live.view.BDCloudVideo.BDCloudVideoView
import com.live.view.toast
import kotlinx.android.synthetic.main.activity_bdlive.*

class BDLiveActivity : BaseActivity(), IMediaPlayer.OnPreparedListener, IMediaPlayer.OnCompletionListener, IMediaPlayer.OnErrorListener, IMediaPlayer.OnInfoListener, IMediaPlayer.OnBufferingUpdateListener, BDCloudVideoView.OnPlayerStateListener {

    private var url: String? = null

    private var bright = 0

    private var mAudioManager: AudioManager? = null

    private var isClickBack = false

    private var startY = 0f//手指按下时的Y坐标
    private var startX = 0f//手指按下时的X坐标


    private val screenBrightness: Int
        get() {
            var screenBrightness = 255
            try {
                screenBrightness = Settings.System.getInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS)
            } catch (ignored: Exception) {

            }

            return screenBrightness
        }

    private var countDownTimer: CountDownTimer? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val window = this.window
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            // finally change the color
            window.statusBarColor = -0xd7d7d8
        }
        //防闪屏
        window.setFormat(PixelFormat.TRANSLUCENT)
        setContentView(R.layout.activity_bdlive)

        url = savedInstanceState?.getString("url") ?: intent.getStringExtra("url")

        bright = screenBrightness
        //        //初始化音频管理器
        mAudioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager

        initPlayer()

        vv_activty_bdlive.setVideoPath(url)
        // 初始化好之后立即播放（您也可以在onPrepared回调中调用该方法）
        if (!vv_activty_bdlive.isPlaying) {
            vv_activty_bdlive.start()
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        outState?.putString("url", url)
        super.onSaveInstanceState(outState)
    }


    private fun initPlayer() {
        //设置ak
        //        BDCloudVideoView.setAK("b98ce254325147ec83572bcf2abfc43d");
        //注册listener
        vv_activty_bdlive.setOnPreparedListener(this)
        vv_activty_bdlive.setOnCompletionListener(this)
        vv_activty_bdlive.setOnErrorListener(this)
        vv_activty_bdlive.setOnInfoListener(this)
        vv_activty_bdlive.setOnBufferingUpdateListener(this)
        vv_activty_bdlive.setOnPlayerStateListener(this)

        vv_activty_bdlive.setVideoScalingMode(BDCloudVideoView.VIDEO_SCALING_MODE_SCALE_TO_FIT)

        vv_activty_bdlive.setLogEnabled(false)
        //        vv_activty_bdlive.setDecodeMode(BDCloudMediaPlayer.DECODE_SW);
        vv_activty_bdlive.setMaxProbeTime(2000) // 设置首次缓冲的最大时长
        vv_activty_bdlive.setTimeoutInUs(1000000)
        // Options for live stream only
        //        vv_activty_bdlive.setMaxProbeSize(1 * 2048);
        //        vv_activty_bdlive.setMaxProbeTime(40); // 设置首次缓冲的最大时长
        //        vv_activty_bdlive.setMaxCacheSizeInBytes(32 * 1024);
        //        vv_activty_bdlive.setBufferTimeInMs(100);
        //        vv_activty_bdlive.toggleFrameChasing(true);
    }

    private fun setLight(brightness: Int) {
        val temp = bright + brightness
        if (temp < 0 || temp > 255) {
            return
        }
        bright = temp
        val lp = window.attributes
        lp.screenBrightness = bright * (1f / 255f)
        window.attributes = lp
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val screenWidth = vv_activty_bdlive.width
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                startX = event.x
                startY = event.y
                if (startY < DeviceUtil.dip2px(50f)) {
                    return super.onTouchEvent(event)
                }
            }
            MotionEvent.ACTION_MOVE -> {
                if (startY < DeviceUtil.dip2px(50f)) {
                    return super.onTouchEvent(event)
                }
                val endY = event.y
                val distanceY = startY - endY
                if (startX > screenWidth / 2) {
                    //右边
                    //在这里处理音量
                    val FLING_MIN_DISTANCE = DeviceUtil.dip2px(3f).toDouble()
                    if (distanceY > FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        mAudioManager!!.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_RAISE, AudioManager.FX_FOCUS_NAVIGATION_UP)
                    }
                    if (distanceY < FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        mAudioManager!!.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_LOWER, AudioManager.FX_FOCUS_NAVIGATION_UP)
                    }
                } else {
                    //屏幕左半部分上滑，亮度变大，下滑，亮度变小
                    val FLING_MIN_DISTANCE = 10.0
                    if (distanceY > FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        setLight(10)
                    }
                    if (distanceY < FLING_MIN_DISTANCE && Math.abs(distanceY) > FLING_MIN_DISTANCE) {
                        setLight(-10)
                    }
                }
                startY = endY
            }
        }
        return super.onTouchEvent(event)
    }


    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (countDownTimer != null) {
            countDownTimer!!.cancel()
        }

        val url = intent.getStringExtra("url")

        if (this.url == url) {
            vv_activty_bdlive.enterForeground()
            this.url = url
        } else {
            vv_activty_bdlive.setVideoPath(url)
            vv_activty_bdlive.start()
        }

    }

    override fun onRestart() {
        super.onRestart()
        if (isClickBack) {
            return
        }
        if (vv_activty_bdlive != null) {
            vv_activty_bdlive.enterForeground()
        }
    }

    override fun onResume() {
        super.onResume()
        isClickBack = false
        startBackground = 2
    }

    override fun onBackPressed() {
        var intent: Intent? = intent
        if (intent == null) {
            intent = Intent(this, AnchorListActivity::class.java)
        } else {
            val componentName = ComponentName(this, AnchorListActivity::class.java)
            intent.component = componentName
        }
        startActivity(intent)
        moveTaskToBack(isTaskRoot)
        isClickBack = true
        if (vv_activty_bdlive != null) {
            vv_activty_bdlive.stopPlayback()
            this.url = ""
            //后台60s自动关闭
            if (countDownTimer == null) {
                countDownTimer = object : CountDownTimer(60000, 30000) {
                    override fun onTick(millisUntilFinished: Long) {

                    }

                    override fun onFinish() {
                        finish()
                    }
                }
            }
            countDownTimer!!.start()
        }
    }


    override fun onStop() {
        // enterBackground should be invoke before super.onStop()
        if (isClickBack) {
            super.onStop()
            return
        }
        if (vv_activty_bdlive != null) {
            vv_activty_bdlive.enterBackground()
        }
        startBackground = 1
        super.onStop()
    }

    override fun onDestroy() {
        if (vv_activty_bdlive != null) {
            vv_activty_bdlive.stopPlayback() // 释放播放器资源
            vv_activty_bdlive.release() // 释放播放器资源和显示资源
        }
        if (countDownTimer != null) {
            countDownTimer!!.cancel()
            countDownTimer = null
        }
        startBackground = 0
        super.onDestroy()
    }


    override fun onInfo(mp: IMediaPlayer, what: Int, extra: Int): Boolean {
        return false
    }

    override fun onError(mp: IMediaPlayer, what: Int, extra: Int): Boolean {
        toast("出错啦出错啦出错啦")
        finish()
        return false
    }

    override fun onCompletion(mp: IMediaPlayer) {

    }

    override fun onPrepared(mp: IMediaPlayer) {

    }

    override fun onBufferingUpdate(mp: IMediaPlayer, percent: Int) {

    }

    override fun onPlayerStateChanged(nowState: BDCloudVideoView.PlayerState) {

    }

    companion object {
        //0 activity未启动 1 activity在后台 2 activity在前台
        var startBackground = 0
    }
}
